### PR TITLE
Gamify, a plugin to gamify your GTG workflow

### PR DESCRIPTION
**Pull request by zeddo123**
Hi,
I have implemented the three main feature that I think are most important for this kind of plugins.

### Features
1. Count the number of tasks done per day.
2. Count the number of days in which the user has completed the goal
3. Added a score and level.
* message and emoji are displayed depending on the number of tasks done.
* The scores are calculated depending on a tag
  * if the task has `@hard` as a tag, the user receives +3 points.
  * if the task has `@meduim` as a tag, the user receives +2 points.
  * if the task has `@easy` as a tag, the user receives +1 point.
* Levels
  * 100 Beginner
  * 1000 Novice
  * 2000 professional
  * 4000 Expert
  * 9000 Master
  * 13000 Master II
  * 19000 Grand Master
  * 25000 Productivity Lord

### Preferences
For the moment, the user can only change one thing which is the goal number of tasks per day.

### Plugin API
I had to add a couple of method to the plugin API so that I get the headerbar.

### Some screenshots
[No task marked as done](https://user-images.githubusercontent.com/34837535/98710585-234f8000-2384-11eb-8bce-9c57c1710fdd.png)
[one or more tasks are done](https://user-images.githubusercontent.com/34837535/98710733-598cff80-2384-11eb-8e33-d5a0f82af4aa.png)
[achieved daily goal](https://user-images.githubusercontent.com/34837535/98710929-91944280-2384-11eb-96ca-2827e05bd688.png)

### Update
* Added the GtkLevelBar to the plugin, as well as the possibility to activate one or the other.
* Ability to add custom tag mappings from the settings of the plugin

Hope you like it.
ref #435 #441

### CODE REVIEW

**Comment by diegogangl** [comment]
Hope you don't mind, I'm setting this for 0.6 for now so we can put 0.5 out soon(ish).

File: GTG/plugins/gamify/gamify.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+        if easy:
+            return 1
+        if meduim:
+            return 2
+        if hard:
+            return 3
+        return 1
+
+    def OnTaskOpened(self, plugin_api):
+        pass
```
Comment (nekohayo) [inline GTG/plugins/gamify/gamify.py:160]: Did you want to do something here and changed your mind, or is the plugin API forcing this method to be present?

File: GTG/plugins/gamify/gamify.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+
+    def button_update_streak(self):
+        """Update the streak numbers in the BUTTON widget"""
+        streak_number = self.builder.get_object('streak_number')
+        streak_emoji = self.builder.get_object('streak_emoji')
+        if self.get_streak() > 0:
+            streak_emoji.set_markup("\U0001F525")
+        else:
+            streak_emoji.set_markup("\U0001F9CA")
+        streak_number.set_markup(_("You've completed your goal {streak} day in a row.").format(streak=self.get_streak()))
```
Comment (nekohayo) [inline GTG/plugins/gamify/gamify.py:293]: You might need to use ngettext (if I recall correctly) to handle plurals for the streak here, for example day vs days... unless you cheat and consider it a streak only if `get_streak() > 1`  :)

**Review by nekohayo — COMMENTED** [review]
I didn't pay close attention to the code, but I can at least say that the plugin seems to work, from some basic testing I've done. It wouldn't shock me to merge and include it for 0.5?

Some additional remarks/questions:
* What is the "analytics" and "preferences" file format? ini? pickle? Might be better to add a filename extension for that
* The plugin doesn't take into account marking tasks as not-done (ex: un-doing/un-dismissing a task), though I don't know if it should / needs to...

**Comment by zeddo123** [comment]
I've added the ability to add custom mappings from the settings, and the possibility for translation.

>What is the "analytics" and "preferences" file format? ini? pickle? Might be better to add a filename extension for that

I'm using the default way to save data of plugins from the , so it should be pickle. Which file extension should I add?

>The plugin doesn't take into account marking tasks as not-done (ex: un-doing/un-dismissing a task), though I don't know if it should / needs to...

We probably should add that feature. However, from the current implementation it won't be perfect as some tag mappings might have been deleted or altered (the deduction of points will be wrong).

>Should the difficulty/points-weighing tags be configurable (or dual-translation-capable like the dates keywords elsewhere in the app)?

That's done now. :)

**Comment by leio** [comment]
A quick drive-by comment on one aspect only:
There appear to be a bunch of new files added, including python source files that have strings marked for translation, but only the plugin metadata file is added to POTFILES.in, therefore nothing else will be translatable (the translatable strings won't be extracted for translators with the i18n tooling).

**Comment by zeddo123** [comment]
> A quick drive-by comment on one aspect only:
> There appear to be a bunch of new files added, including python source files that have strings marked for translation, but only the plugin metadata file is added to POTFILES.in, therefore nothing else will be translatable (the translatable strings won't be extracted for translators with the i18n tooling).

Thanks, I'll look into it :D

**Comment by zeddo123** [comment]
After rebasing my branch on master, I got this message saying that I need the `GTG.core.logger` in the plugins window.
image

*I will verify on my laptop, it could be a problem in my machine*

**Comment by diegogangl** [comment]
The logger module got nuked a while back:
https://github.com/getting-things-gnome/gtg/pull/539

**Comment by zeddo123** [comment]
Forgot about that. okay I'll switch to the new logger.

**Comment by diegogangl** [comment]
I've finally sat down to review this! Code looks good, though it could use some enters to separete lines more (and make it more readable). Specially after docstrings.

Also, congrats on being the first to use `:=` in GTG o/

Something I noted is that the score doesn't update when tasks are marked as not done. A cheater could use this to increase his score without actually doing the tasks.

## Popup

I would add some more padding between the large emoji and text of the first section. Don't be afraid to play with font size to improve the hierarchy of information.

"your current level" can be simplified to "level:" and points can be put below without any other text. Like this:

---

Level: **Beginner**
5 Points

---

The streak can be simplified to a single line like this (emoji and text in the same line):

---

🔥️ **2 day** streak
👻️ First day completing your goals (couldn't find the ice cube emoji for some reason)

---


## Preferences

I dig the fancy title at the top!

- I would move "Select UI Version" to below "Target Tasks per day", and make a listbox to hold both settings so they look like the main preferences. Look at `preferences.ui` for how that's done in the preferences dialog.

- You can set the "Target Tasks per day" value to zero and hit your goal every day forever :)

- Same thing for the tags section. Look at the `preferences.ui` file as well as the `style.css` file (I had to add some classes). Instead of the `Add` and `Delete` buttons, add a new row at the end with a + to add a new tag, and a button to each row to delete that row. Look at the "Region & Language" section of GNOME's settings for an example.

- Watch out for the @ in the text input, so the user doesn't end up adding `@@tag`.

- "Select UI Version" can be simplified to "UI Mode". I also think a [Combo Box](https://lazka.github.io/pgi-docs/index.html#Gtk-3.0/classes/ComboBox.html#example) would be a better fit than the radio buttons.

## Future ideas

- The difficulty tags are cool, but in the future there could be some heuristics to calculate score like number of subtasks, time a task has been open, number of swear words, etc.
- It would be cool to include some messages that encourage people to relax once they hit their goal (for workaholics/people stressed out of their minds)

**Comment by zeddo123** [comment]
I've been working on your proposed enhancement. Just one thing though.

>Something I noted is that the score doesn't update when tasks are marked as not done. A cheater could use this to increase his score without actually doing the tasks.

This brings the problem of negative number of tasks done. Let say, today I only marked a task as not done. if I implement this feature the easy way, you would have a negative number of points. What would you think would be the best thing to do here?

We could either:
* Do it the easy way (we enable points to be negative)
* Only deduce points when the task has been marked as done the same day.

**Comment by diegogangl** [comment]
> This brings the problem of negative number of tasks done. Let say, today I only marked a task as not done. if I implement this feature the easy way, you would have a negative number of points. @diegogangl What would you think would be the best thing to do here?
> 
> We could either:
> 
>     Do it the easy way (we enable points to be negative) or limit it to drop only to 0.
>     Only deduce points when the task has been marked as done the same day.

I think the simplest would be to clamp the score to 0.

**Comment by zeddo123** [comment]
Hey @diegogangl. I've made the necessary changes. Whenever you have time, take a look.

**Comment by diegogangl** [comment]
Hey o/
The popup looks great. I like the new icons you added (specially the star!).
I still have some changes for the preferences:

- Drop the sidebar. Sorry I wasn't clear on this, I meant the general preferences page. We're also going to drop that sidebar in the apps preferences in the future (once we get to libhandy/libadwaita)
- Use `user-trash-symbolic` as the icon for _Delete Tag_ button
- Add the `.prefs_list` css class to the listboxes so they get borders and better padding
- Remove the "close" button, it duplicates the X button
- Push the apply button to the other side of the window

**Comment by zeddo123** [comment]
Yeah it looks much better now! Should I squash all commits into one?

**Comment by diegogangl** [comment]
Looking good!
I would squash the last three commits into one, and leave the rest of the history.

There's still one a couple of things left:
- Make the labels left-aligned
- There's no need to repeat UI in the menu when it's already in the label. I would change "Button UI" to "Button only" and "Full/Both" to "Button and Progress Bar", and "Discrete" to "Progress bar"
- The text in the tags labels should be centered vertically

Last changes! I would also squash this last commit with the other 3

**Comment by zeddo123** [comment]
>The text in the tags labels should be centred vertically

Do you want them to be in the centre of the area, and not left-aligned?

**Comment by diegogangl** [comment]
> > The text in the tags labels should be centred vertically
> 
> Do you want them to be in the centre of the area, and not left-aligned?

Left-aligned horizontally, centered vertically. (I think it's alignment and padding in Glade, 0 for horizontal and 0.5 for vertical)

**Comment by zeddo123** [comment]
done :smile:

**Comment by diegogangl** [comment]
It's been a long road, thanks for the work!

**Pull request merged**